man/man1/kiinfo.1
.\" Process this file with
.\" groff -man -Tascii kiinfo.1
.\"
.ad
.TH kiinfo 1 "7.11 - May 06, 2025" version "7.11"
.SH NAME
kiinfo  -  LinuxKI data collection and analysis engine

.SH SYNOPSIS
.B kiinfo \-<kitool> [FLAGS] [OPTIONS]

.SH DESCRIPTION
\fBkiinfo\fR is one of the primary components of the LinuxKI Toolset. It is used to enable and dump the trace data using either ftrace or the LiKI tracing mechanisms and it is also used to analyze the trace data from either a LinuxKI dump collected with the runki script or from a running system.

\fBkiinfo\fR is executed by both the \fBrunki\fR script to collect a LinuxKI dump as well as from the \fBkiall\fR script to generate the LinuxKI reports as part of the post-processing. Most of the time, using the runki and kiall scripts is sufficient. However, there are times when a more customized collection or a customized report is needed.

\fBkiinfo\fR works on LinuxKI dump files using the -ts <timestamp> option, as well as live systems using the -a <secs> and -p <passes> options. Kiinfo also requires one of the subtools <kitool> listed in the Options, such as -kitrace.   See the Examples below for additional details.   


.SH OPTIONS

.B \-h | \-help
.RS 
Print kiinfo syntax
.RE

.B Options for KI Dump analysis

.B \-ts | \-timestamp <timestamp>
.RS 
Used when analyzing LinuxKI dumps.   Timestamp is from ki.all.<timestamp>.tgz file created by runki script.  Cannot be used with -a option.
.RE

.B \-start | \-starttime <time_in_secs)
.RS 
Report only data after the start time.   Only used with KI dumps (-ts option)
.RE

.B \-end | \-endtime <time_in_secs)
.RS 
Report only data before the end time.   Only used with KI dumps (-ts option)
.RE

.B Options for online analysis

.B \-a | \-alarm <seconds>
.RS 
Length of interval for online KI analysis using the LiKI tracing mechanism (default 5).  Cannot be used with -ts option.
.RE

.B \-p | \-passes <passes>
.RS 
Number of intervals/passes for online analysis (default 1).  Used with the -a option.
.RE

For online analysis, be sure the debug filesystem (debugfs) is mounted:

.B $ mount -t debugfs debugfs /sys/kernel/debug 

.B Kiinfo has a number of tool options.  Choose one of the following tool options to execute.

.B \-dump | \-kitracedump 
.RS 
Enable Ftrace and dump debugfs per-cpu ring buffer data to disk. See kiinfo-kitracedump(1).
.RE

.B \-likid | \-likidump 
.RS 
Enable LiKI and dump debugfs per-cpu ring buffer data to disk. Assumes likit.ko DLKM module is loaded. See kiinfo-likidump(1). 
.RE

.B \-likim | \-likimerge
.RS
Merge ki.bin.<cpu>.<timetamp> files into a single ki.bin file (LiKI only). See kiinfo-likimerge(1). 
.RE

.B \-live [FLAGS]
.RS
curses-based user interface. See kiinfo-live(1).
.RE

.B \-kparse | \-kp [FLAGS]
.RS
Kparse System Overview report - first pass analysis tool
.RE

.B \-kitrace [FLAGS]
.RS 
KI ASCII trace format tool. See kiinfo-kitrace(1).
.RE

.B \-kipid [FLAGS]
.RS 
KI PID Analysis Report. See kiinfo-kipid(1).
.RE

.B \-kiprof [FLAGS]
.RS 
KI CPU Profiling Report. See kiinfo-kiprof(1)
.RE

.B \-kidsk [FLAGS]
.RS 
KI Disk Analysis Report. See kiinfo-kidsk(1).
.RE

.B \-kirunq [FLAGS]
.RS 
KI CPU/RUNQ Analysis Report. See kiinfo-kirunq(1).
.RE

.B \-kiwait [FLAGS]
.RS 
KI Wait Event Analysis Report. See kiinfo-kiwait(1).
.RE

.B \-kifile [FLAGS]
.RS 
KI File Activity Report. See kiinfo-kifile(1).
.RE

.B \-kisock [FLAGS]
.RS 
KI Network Socket Activity eeport. See kiinfo-kisock(1).
.RE

.B \-kifutex [FLAGS]
.RS 
KI Futex Activity Report. See kiinfo-kifute(1).
.RE

.B \-kidock [FLAGS]
.RS 
KI Docker Activity Report. See kiinfo-kidock(1).
.RE

.B \-kiall [FLAGS]
.RS 
Generate all KI reports in a single pass (except ASCII trace data).  Typically invoked by kiall script. See kiinfo-kiall(1).
.RE

.B \-clparse [FLAGS]
.RS 
KI Cluster Report.  Reports on several KI trace collections from difference nodes in a cluster.  Each KI trace collection must have the same. See kiinfo-clparse(1).
.RE

.SH EXAMPLES

1) Collect LinuxKI trace data using ftrace for 30 seconds and provide a "tag" 
using the current timestamp:

.B $ kiinfo -kitracedump dur=30 -ts `date +%m%d_%H%M`

2) Get kparse html output from files collected by runki

.B $ kiinfo -kparse -html -ts 0901_1548 >kp.html

3) Create the Disk Analysis report and also show top 10 tasks performing disk I/O

.B $ kiinfo -kidsk npid=10 -ts 0901_1548

4) Analyze disk with major=0x68 and minor=0x0 (0x68,0) on live system and show top 5 processes accessing the device. Show results every 5 seconds for 6 passes.

.B $ kiinfo -kidsk dev=0x06800000,npid=5 -a 5 -p 6

5) Trace a live process for 2 seconds

.B $ kiinfo -kitrace pid=5836 -a 2 >pid.5836

6) Generate a Cluster Overview Report for several nodes in a cluster with the same timestamp:

.B $ kiinfo -clparse csv,cltree,top=20 -html -ts 0422_1517

7) Get help information on the -kipid option:

.B $ kiinfo -kipid help

8) Start curses-based user interface with 5 second (default) updates

.B $ kiinfo -live 

.SH AUTHOR
Mark C. Ray <mark.ray@hpe.com>

.SH SEE ALSO
LinuxKI(1) kiinfo-dump(1) kiinfo-likidump(1) kiinfo-likimerge(1) kiinfo-live(1) kiinfo-kparse(1) kiinfo-kitrace(1) kiinfo-kipid(1) kiinfo-kiprof(1) kiinfo-kidsk(1) kiinfo-kirunq(1) kiinfo-kiwait(1) kiinfo-kifile(1) kiinfo-kisock(1) kiinfo-kifutex(1) kiinfo-kidock(1) kiinfo-kiall(1) kiinfo-clparse(1) runki(1) kiall(1) kiclean(1) kivis-build(1) kivis-start(1) kivis-stop(1)

https://github.com/HewlettPackard/LinuxKI/wiki



src/kiinfo/kiinfo.c
/***************************************************************************
 * Copyright 2017 Hewlett Packard Enterprise Development LP.  
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version. This program is distributed in the
 * hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU General Public License for more details. You
 * should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * ***************************************************************************/

#include <strings.h>
#include <string.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/utsname.h>
#include <pwd.h>
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include "ki_tool.h"
#include "option_iface.h"
#include "liki.h"
#include "winki.h"
#include "globals.h"
#include "developers.h"
#include "info.h"
#include "hash.h"
#include "vers.h"
#include "html.h"

extern void (*init_func)(void *);
extern void firstpass_init_func(void *);
extern int open_trace_files();
extern int close_trace_files(int);
extern int init_trace_files(int);
extern int open_merged_file();
extern int init_debug_mountpoint(char *);

int debug=0;            /* debug code is off by default */
int is_alive=0;         /* live analyze can be done with liki kernel module only */
int VM_guest=0;         /* for live analysis, is this system a VM Guest */

char arch_flag = X86_64;
struct tm gmt;
char *cmdstr;
char *timestamp;
int do_detail_help;
int alarm_secs=5;               
char HTML	= 0;
int passes=1;
int done=FALSE;
init_t Init;
char startup_found = 0;
double realsecs = 0;
uint64 realtime = 0;
uint64 start_time = 0;				/* absolute times of first event */
uint64 winki_start_time = 0;			/* absolute times of first traced event */
uint64 end_time= 0;
uint64 prev_vint_time = 0;
uint64 vis_hostid = 0;
uint64 interval_start = 0;			/* VIS csv interval reporting */
uint64 interval_end = 0;			/* VIS csv interval reporting */
struct timespec begin_time;			
int64 start_filter = 0;			/* relative times */
int64 start_filter_save;
int64 end_filter= 0xfffffffffffffffull;
int64 end_filter_save;
int64 kistep = 0ull;
uint32 winki_bufsz = 0;
char	liki_module_loaded = 0;
char	liki_initialized = 0;
char *cwd = NULL;
int nfiles = 0;
struct utsname utsname;
char *kgdboc_str = NULL;
char *tlabel;		/* Thread label for reporting Threads ID */
char *plabel;		/* Process/Task Group label for reporting Process ID */
EventTraceHdr_t *winki_hdr = NULL;

trace_info_t  trace_files[MAXCPUS];
trace_info_t  trace_file_merged;
char ts_begin_marker[64];
char ts_end_marker[64];

FILE *dockfile = NULL;
FILE *node_csvfile = NULL;
FILE *pid_csvfile = NULL;
FILE *dsk_csvfile = NULL;
FILE *futex_csvfile = NULL;
FILE *file_csvfile = NULL;
FILE *socket_csvfile = NULL;
FILE *runq_csvfile = NULL;
FILE *wait_csvfile = NULL;
FILE *cluster_csvfile = NULL;
FILE *cluster_network_csvfile = NULL;
FILE *server_vis_csvfile = NULL;
FILE *cluster_vis_csvfile = NULL;
FILE *syscofig_file = NULL;

#include "custom_options.h"


static void catch_int(int sig)
{
	fprintf (stderr, "kiinfo caught Signal %d\n", sig);
}

int
main(int argc, char *argv[])
{
	char	command[4096];
	int i, n, ret;
	char line[256];
	char kp[30];
	FILE *trcfile = NULL;	

	Init.i_argc = argc;
	Init.i_argv = argv;
	Init.i_options = otab;
	cmdstr = argv[0];

	/* output toolname and arguments */
	sprintf (command, "Command line: %s ", argv[0]);
	for (i=1; i<argc; i++) {
		n = strlen(command);
		sprintf(&command[n], "%s ", argv[i]);
	}

	option_decode(&Init);  
	if (timestamp == NULL) is_alive = TRUE;

	if (kiall_flag) {
		if (timestamp == NULL) {
			option_usage(&Init, "Timestamp (-ts <timestamp>) required to for kiall tool", "kiall");
			_exit(1);
		} else if (HTML) {
			sprintf(kp, "kp.%s.html", timestamp);
		} else {
			sprintf(kp, "kp.%s.txt", timestamp);
		}
        	if (freopen(kp, "w", stdout) == NULL) {
			fprintf (stderr, "Unable to reopen stdout as %s (errno %d)\n", kp, errno);
			_exit(1);
		}
	}

	if (!kilive) {
		DOC_START(tool_name);
	}

	sprintf (line, "%s %s ", tool_name, timestamp); TITLE(line);
	if (!kilive) {
		STYLE;
		printf("%s\n\n", command);
		BR;
	}

	sprintf (line, "%s (%s)\n", tool_name, tool_version); H1(line); 
	TEXTx("\n");
	HR; BR;
	PRE;

	if (likidump_flag || likistart_flag) {
		if (geteuid() != 0) { 
			fprintf (stderr, "You must run kinfo as root to collect a trace dump\n");
			_exit(-1);
		}

		if (uname(&utsname) == 0) {
			if (strstr(utsname.machine, "aarch64")) arch_flag = AARCH64;
			else if (strstr(utsname.machine, "ppc64le")) arch_flag = PPC64LE;
		}
		clear_kgdboc();
		syscall_arg_list = linux_syscall_arg_list;

		likidump();
		if (kgdboc_str) reset_kgdboc();
		_exit(0);
	} else if (likiend_flag) {
		if (geteuid() != 0) { 
			fprintf (stderr, "You must run kinfo as root to collect a trace dump\n");
			_exit(-1);
		}
		
		likiend();

		_exit(0);
	} else if (kitracedump_flag) {
		if (geteuid() != 0) { 
			fprintf (stderr, "You must run kinfo as root to collect a trace dump\n");
			_exit(-1);
		}
		kitracedump();
		_exit(0);
	} else if (objdump_flag) {
		objdump();
		_exit(0);
	} else if (etldump_flag) {
		etldump();
		_exit(0);
	}

	if (likimerge_flag) {
		if (timestamp == NULL) {
			option_usage(&Init, "Timestamp (-ts <timestamp>) required to merge KI binary files", "likimerge");
			_exit(1);
		}
		globals = GET_SERVER(server[0]);
		globals->server_id = nservers;
		nfiles = open_trace_files();
        	if (nfiles == 0) {
			FATAL(1008, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
        	}

		init_trace_files(nfiles);		
		close_trace_files(nfiles);		
		merge();
	}

	if (kitracedump_flag || likimerge_flag) {
		_exit(0);
	}

	if (tool_init_func == NULL) _exit(0);

	if (is_alive) CLEAR(COOP_DETAIL_ENABLED);
	if (is_alive && kparse_flag) {
		fprintf(stderr, "Live tracing not allowed with kiinfo -kparse\n");
		_exit(1);
	}

	if (csv_flag) fsep = ',';               /* change field separator for CSV output */ 
	if (is_alive) {
		if (geteuid() != 0) { 
			fprintf (stderr, "You must run kinfo as root to perform online tracing\n");
			_exit(-1);
		}
		/* clear kgdboc file to avoid failure due to jprobe */
		clear_kgdboc();
		/* Disable time filters for live KI */
		start_filter = 0;
		end_filter = 0xfffffffffffffffull;


		if (uname(&utsname) == 0) {
			if (strstr(utsname.machine, "aarch64")) arch_flag = AARCH64;
			else if (strstr(utsname.machine, "ppc64le")) arch_flag = PPC64LE;
		}

		globals = GET_SERVER(server[0]);
		globals->server_id = nservers;
		globals->VM_guest = VM_guest;
		parse_uname(0);
		nservers++;

		syscall_arg_list = linux_syscall_arg_list;

		if (arch_flag == AARCH64) {
			globals->syscall_index_32 = syscall_index_aarch_64;
			globals->syscall_index_64 = syscall_index_aarch_64;
		} else if (arch_flag == PPC64LE) {
			globals->syscall_index_32 = syscall_index_ppc64le;
			globals->syscall_index_64 = syscall_index_ppc64le;
		} else if (IS_WINKI) {
			globals->syscall_index_32 = syscall_index_win;
			globals->syscall_index_64 = syscall_index_win;
		} else {
			globals->syscall_index_32 = syscall_index_x86_32;
			globals->syscall_index_64 = syscall_index_x86_64;
		}

		ki_actions = liki_action();
		start_time = 0;
		end_time= 0;
		globals->kiversion = TRACE_VERSION;		/* initialize this to 1 for now until we actually read data */
		init_trace_ids();
		load_liki_module();

		/* open debugfs directory needef or LiKI */
		init_debug_mountpoint(debug_dir);

		tlabel="PID";
		plabel="TGID";

		init_func(NULL);
		developers_init();		/* calls liki_init_tracing */ 

		set_ioflags(); 	/* needs to be called after setting IS_LIKI */
		set_gfpflags();
	 
		read_liki_traces();		/* call instead of developers_call() */
		liki_close_live_stream();
		unload_liki_module();

		if (kgdboc_str) reset_kgdboc();

		if (!kilive) {
			save_and_clear_server_stats(nfiles);
			printf ("\nTotal time captured (secs): %3.2f\n", globals->total_secs);
        		print_cpu_buf_info();
		} else {
			live_cleanup_func(NULL);
		}
	} else if (timestamp && kilive) {
		cwd = get_current_dir_name();
		for_each_file(".", "ki.bin.", timestamp, cluster_flag);
		if (nservers == 0) {
			FATAL(1009, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
		}

		sprintf (ts_begin_marker, "kitrace_marker_BEGIN_%s", timestamp);
		sprintf (ts_end_marker, "kitrace_marker_END_%s", timestamp);
		
		globals = server[0];
		start_time = 0;
		end_time= 0;

		nfiles = open_merged_file();
        	if (nfiles == 0) {
			nfiles = open_trace_files();
		}

		if (nfiles == 0) FATAL(1010, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);

		if (IS_WINKI) {
			tlabel="TID";
			plabel="PID";
		} else {
			tlabel="PID";
			plabel="TGID";
		}

		init_trace_files(nfiles); 
		if (!IS_LIKI) kistep = 0;
		if (IS_FTRACE) read_fmt_files(); 
		if (!IS_WINKI) init_trace_ids(); /* skip this for now */

		parse_uname(1);
		set_ioflags();
		set_gfpflags();
		printf ("KI Binary Version %d\n", globals->kiversion);

		if (IS_WINKI) {
			syscall_arg_list = win_syscall_arg_list;
		} else {
			syscall_arg_list = linux_syscall_arg_list;
                }

		if (arch_flag == AARCH64) {
			globals->syscall_index_32 = syscall_index_aarch_64;
			globals->syscall_index_64 = syscall_index_aarch_64;
		} else if (arch_flag == PPC64LE) {
			globals->syscall_index_32 = syscall_index_ppc64le;
			globals->syscall_index_64 = syscall_index_ppc64le;
		} else if (IS_WINKI) {
			globals->syscall_index_32 = syscall_index_win;
			globals->syscall_index_64 = syscall_index_win;
		} else {
			globals->syscall_index_32 = syscall_index_x86_32;
			globals->syscall_index_64 = syscall_index_x86_64;
		}

		if (IS_WINKI) {
			int reset_kitrace_flag = 0;
			
			firstpass_init_func(NULL);
			developers_call(nfiles);
			close_trace_files(nfiles);

			nfiles = open_trace_files();
			if (nfiles == 0)  FATAL(1010, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
			init_trace_files(nfiles);

			start_time = 0;
		} else {
			end_filter_save = end_filter;
			start_filter_save = start_filter;
			if (kistep) end_filter = start_filter + kistep;
		}

		if (IS_WINKI) {
			tlabel="TID";
			plabel="PID";
		} else {
			tlabel="PID";
			plabel="TGID";
		}

		init_func(NULL);

		while (!done) {
			developers_call(nfiles);
			developers_report();
			clear_all_stats();
			reset_trace_files(nfiles);		
			start_time = 0;
			end_time= 0;
			init_trace_files(nfiles); 
			globals->missed_events = globals->total_events = 0;
                        parse_pself();
                        parse_proc_cgroup();
		}

		close_trace_files(nfiles);		
		live_cleanup_func(NULL);
		
	} else if (timestamp) {
		/* Ignoring system calls is applied during data collection only */
		if (sysignore) { 
			fprintf (stderr, "Warning: sysignore file is only applied at the time the trace is collected.  Ignoring sysignore option.\n");
		}

		cwd = get_current_dir_name();
		for_each_file(".", "ki.bin.", timestamp, cluster_flag);

		if (nservers == 0) {
			FATAL(1009, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
		}

		sprintf (ts_begin_marker, "kitrace_marker_BEGIN_%s", timestamp);
		sprintf (ts_end_marker, "kitrace_marker_END_%s", timestamp);
		if (cluster_flag)  
			fprintf (stderr, "Number of Servers to analyze: %d\n", nservers);

		for (i=0; i < nservers; i++) {
			globals = server[i];
			start_time = 0;
			end_time= 0;
			ret = chdir(globals->subdir);
			if (cluster_flag) fprintf (stderr, "Processing KI files in %s\n", globals->subdir);
			nfiles = open_merged_file();
       			if (nfiles == 0) {
				nfiles = open_trace_files();
			}

			if (nfiles == 0)  FATAL(1010, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
			init_trace_files(nfiles); 


			if (IS_FTRACE) read_fmt_files(); 
			if (!IS_WINKI) init_trace_ids();  /* skip for now */

			parse_uname(1);
			set_ioflags();
			set_gfpflags();
			printf ("KI Binary Version %d\n", globals->kiversion);

			if (IS_WINKI) {
				syscall_arg_list = win_syscall_arg_list;
			} else {
				syscall_arg_list = linux_syscall_arg_list;
               		}
			if (arch_flag == AARCH64) {
				globals->syscall_index_32 = syscall_index_aarch_64;
				globals->syscall_index_64 = syscall_index_aarch_64;
			} else if (arch_flag == PPC64LE) {
				globals->syscall_index_32 = syscall_index_ppc64le;
				globals->syscall_index_64 = syscall_index_ppc64le;
			} else if (IS_WINKI) {
				globals->syscall_index_32 = syscall_index_win;
				globals->syscall_index_64 = syscall_index_win;;
			} else {
				globals->syscall_index_32 = syscall_index_x86_32;
				globals->syscall_index_64 = syscall_index_x86_64;
			}

			if (IS_WINKI) {
				int reset_kitrace_flag = 0;
			
				if (kitrace_flag) {
					CLEAR(KITRACE_FLAG);
					reset_kitrace_flag=1;
				}
				firstpass_init_func(NULL);
				developers_call(nfiles);
				close_trace_files(nfiles);

				nfiles = open_trace_files();
				if (nfiles == 0)  FATAL(1010, "Cannot open file of the form ki.bin*.<timestamp>", timestamp, -1);
				init_trace_files(nfiles);
				if (reset_kitrace_flag) {
					SET(KITRACE_FLAG);
				}
			}

			start_time = 0;

			if (IS_WINKI) {
				tlabel="TID";
				plabel="PID";
			} else {
				tlabel="PID";
				plabel="TGID";
			}

			init_func(NULL);

			developers_init();
			developers_call(nfiles);
			close_trace_files(nfiles);		
			save_and_clear_server_stats(nfiles);
			ret = chdir(cwd);
		}

		developers_report();

		if ((!cluster_flag) && (!kiall_flag) && (!kilive)) {
			printf ("\nTotal time captured (secs): %3.2f\n", globals->total_secs);
        		print_cpu_buf_info();
		}

		if (kilive) {
			live_cleanup_func(NULL);
		}
	}

	exit(0);
}
